Add `intersect(::AbstractRange, ::AbstractVector)`

barucden opened the pull request:
Closes #41759

barucden committed:
Add `intersect(::AbstractRange, ::AbstractVector)`

Also adds:
 - `intersect(::AbstractRange, ::AbstractRange)`
 - `intersect(::AbstractRange, ::AbstractRange)`

Closes #41759

Co-authored-by: Ian Butterworth <i.r.butterworth@gmail.com>
Co-authored-by: Jeff Bezanson <jeff.bezanson@gmail.com>

barucden commented:
@IanButterworth Cannot refer to `AbstractVector` from `range.jl`. Should I just move the definitions to `array.jl` (originally I thought it is more fitting in `range.jl`)?

IanButterworth commented:
Moving to `array.jl` sounds good. Perhaps stylistically you could flip the args so the first arg makes most sense for the new location, but that's very minor
```julia
function intersect(vec::AbstractVector, r::AbstractRange)
    common = Iterators.filter(x -> x ∈ r, vec)
    seen = Set{eltype(vec)}(common)
    return vectorfilter(_shrink_filter!(seen), common)
end
intersect(r::AbstractRange, vec::AbstractVector) = intersect(vec, r)
```

barucden commented:
Done. Thanks!

IanButterworth commented:
Just copying the benchmarks over from #41759 

```

julia> @btime intersect(1:typemax(Int), [1,3])
  302.226 ns (8 allocations: 688 bytes)
2-element Vector{Int64}:
 1
 3

julia> @btime Base.intersect([1,2,3,4,5], [1,3])
  654.689 ns (15 allocations: 1.11 KiB)
2-element Vector{Int64}:
 1
 3

julia> @btime Base.intersect(1:typemax(Int), 1:2:3)
  0.044 ns (0 allocations: 0 bytes)
1:2:3

````

simeonschaub commented:
We don't currently define an explicit method for `intersect(::AbstractRange, ::AbstractRange)`, so you will need to define that as well to avoid ambiguities.

oscardssmith commented:
Something interesting is that this could be about 2x faster if we used a Set implimentation closer to the one in Dictionaries.jl's. specifically of you use the version of a set that uses insertion order, this could just return the elements of the set as a list without an extra allocation.

barucden commented:
> We don't currently define an explicit method for `intersect(::AbstractRange, ::AbstractRange)`, so you will need to define that as well to avoid ambiguities.

@simeonschaub I don’t immediately see what would be the approach. Should it be just a fall-back method with a naive implementation? Like looping through the shorter range and checking if each element is in the longer range? Also, what ambiguities can occur now that could not occur before this method?


> Something interesting is that this could be about 2x faster if we used a Set implimentation closer to the one in Dictionaries.jl's. specifically of you use the version of a set that uses insertion order, this could just return the elements of the set as a list without an extra allocation.

@oscardssmith Thanks for the suggestion. Is there a set implementation which guarantees the insertion order in Julia? I could not find it. Or do you propose to write it first (maybe based on the mentioned Dictionaries.jl package)?

simeonschaub commented:
> Should it be just a fall-back method with a naive implementation? Like looping through the shorter range and checking if each element is in the longer range?

Yes, seems reasonable. 

> Is there a set implementation which guarantees the insertion order in Julia?

No, not currently. That would be a much bigger project, so I wouldn't bother for now.

DilumAluthge commented:
Can you rebase this on the latest `master` to see if that fixes CI?

You can ignore the failures on `package_freebsd64`. But the other jobs should be passing.

IanButterworth commented:
@DilumAluthge Aren't the (non-freebsd) errors here real ambiguity failures related to this PR?

DilumAluthge commented:
> @DilumAluthge Aren't the (non-freebsd) errors here real ambiguity test failures related to this PR?

Oh you're right. I just assumed they were the Linux `EAI_AGAIN` errors 😂

barucden commented:
Rebased and added `intersect(::AbstractRange, ::AbstractRange)`. @simeonschaub Would you check if [this test](https://github.com/JuliaLang/julia/pull/41769/commits/8c2c5237f4fdba3c5269ee725313d0efd2c365c3#diff-1d6e3ecff651f017ed32b54d0919e13362ac278dd18df1cc8b2feda52e630744) is okay, please?

simeonschaub commented on base/range.jl:
```diff
@@ -1143,6 +1143,13 @@ function intersect(r::StepRange, s::StepRange)
+    if length(r2) < length(r1)
+        r1, r2 = r2, r1
+    end
+    return [x for x in r1 if x ∈ r2]
```
There are range types like `StepRangeLen` which can have step zero and contain more than one element. So we still need some way of ensuring every element in the returned array is unique.

barucden replied:
Is it enough to just do 
```julia
return unique!([x for x in r1 if x ∈ r2])
```
?

simeonschaub replied:
In that case, you can probably just use the approach you used for the other cases already, which should allocate less.

barucden replied:
I changed the test to check the uniqueness too.

mcabbott replied:
Calling `unique` on the ranges might be another way, `unique(StepRangeLen(1,0,10))` seems to do the right thing.

barucden commented:
There's still one more error but I am not sure it's related.

mcabbott commented on base/range.jl:
```diff
@@ -1143,6 +1143,15 @@ function intersect(r::StepRange, s::StepRange)
     step(r) < zero(step(r)) ? StepRange{T,S}(n, -a, m) : StepRange{T,S}(m, a, n)
 end
 
+function intersect(r1::AbstractRange, r2::AbstractRange)
+    if length(r2) < length(r1)
+        r1, r2 = r2, r1
+    end
```
I think this may cause type-stability problems, e.g. `@code_warntype intersect(1:10, 1:1.0:9)`. 

Even if I replace the swop with `return intersect_new(r2, r1)`, I still get `Body::Union{Vector{Float64}, Vector{Int64}}` for this example, due to differing eltypes.

mcabbott replied:
Maybe? The present behaviour BTW is this:
```
julia> intersect(1:1.0:9, 1:3)
3-element Vector{Float64}:
 1.0
 2.0
 3.0

julia> intersect(1:3, 1:1.0:9)
3-element Vector{Int64}:
 1
 2
 3
```
Are there situations where it might matter not to promote? That doesn't sound to me like it ought to be guaranteed by the verb `intersect`.

IanButterworth commented:
@simeonschaub @oscardssmith is this RTM?

oscardssmith commented:
No objections here.

IanButterworth commented:
Some return type comparison and benchmarks vs. master
```julia
using BenchmarkTools
for a in [1:4, 1:1.0:4, 10:4:30, collect(1:4)]
    for b in [2:4, 2:1.0:4, 14:4:30, collect(2:6)]
        r = Base.intersect(a, b)
        print(typeof(a), "\t", typeof(b), "\t", typeof(r),"\t", r, "\t")
        @btime Base.intersect($a, $b)
    end
end
```


| a         | b           | Master return   | PR Return  | Master                                  | PR                                     |
|-----------|-------------|-----------------|-----------------|-----------------------------------------|----------------------------------------|
| 1:4       | 2:4         | 2:4             | 2:4             | 0.010 ns (0 allocations: 0 bytes)       | 0.010 ns (0 allocations: 0 bytes)      |
| 1:4       | 2:1.0:4     | [2, 3, 4]       | [2.0, 3.0, 4.0] | 376.512 ns (16 allocations: 1.12 KiB)   | 66.507 ns (2 allocations: 144 bytes)   |
| 1:4       | 14:4:30     | 14:4:13         | 14:4:13         | 0.010 ns (0 allocations: 0 bytes)       | 0.010 ns (0 allocations: 0 bytes)      |
| 1:4       | [2,3,4,5,6] | [2, 3, 4]       | [2, 3, 4]       | 364.560 ns (14 allocations: 1.03 KiB)   | 156.247 ns (7 allocations: 608 bytes)  |
| 1:1.0:4   | 2:4         | [2.0, 3.0, 4.0] | [2.0, 3.0, 4.0] | 422.495 ns (16 allocations: 1.16 KiB)   | 72.938 ns (2 allocations: 144 bytes)   |
| 1:1.0:4   | 2:1.0:4     | [2.0, 3.0, 4.0] | [2.0, 3.0, 4.0] | 416.785 ns (15 allocations: 1.12 KiB)   | 87.419 ns (2 allocations: 144 bytes)   |
| 1:1.0:4   | 14:4:30     | Float64[]       | Float64[]       | 406.565 ns (15 allocations: 1.08 KiB)   | 44.098 ns (1 allocation: 64 bytes)     |
| 1:1.0:4   | [2,3,4,5,6] | [2.0, 3.0, 4.0] | [2, 3, 4]       | 421.245 ns (14 allocations: 1.06 KiB)   | 234.930 ns (7 allocations: 608 bytes)  |
| 10:4:30   | 2:4         | 10:4:9          | 10:4:9          | 0.010 ns (0 allocations: 0 bytes)       | 0.010 ns (0 allocations: 0 bytes)      |
| 10:4:30   | 2:1.0:4     | Int64[]         | Float64[]       | 401.403 ns (16 allocations: 1.12 KiB)   | 41.474 ns (1 allocation: 64 bytes)     |
| 10:4:30   | 14:4:30     | 14:4:30         | 14:4:30         | 23.391 ns (0 allocations: 0 bytes)      | 21.577 ns (0 allocations: 0 bytes)     |
| 10:4:30   | [2,3,4,5,6] | Int64[]         | Int64[]         | 402.547 ns (14 allocations: 1.03 KiB)   | 88.290 ns (5 allocations: 464 bytes)   |
| [1,2,3,4] | 2:4         | [2, 3, 4]       | [2, 3, 4]       | 335.186 ns (13 allocations: 976 bytes)  | 157.491 ns (7 allocations: 608 bytes)  |
| [1,2,3,4] | 2:1.0:4     | [2, 3, 4]       | [2, 3, 4]       | 345.344 ns (13 allocations: 1008 bytes) | 228.194 ns (7 allocations: 608 bytes)  |
| [1,2,3,4] | 14:4:30     | Int64[]         | Int64[]         | 332.333 ns (13 allocations: 976 bytes)  | 84.778 ns (5 allocations: 464 bytes)   |
| [1,2,3,4] | [2,3,4,5,6] | [2, 3, 4]       | [2, 3, 4]       | 333.701 ns (12 allocations: 944 bytes)  | 341.855 ns (12 allocations: 944 bytes) |

I didn't expect this PR to speed this up so much!

```
# master
julia> @btime intersect(1:1.0:4, 2:4)
  407.815 ns (16 allocations: 1.16 KiB)
3-element Vector{Float64}:
 2.0
 3.0
 4.0

# PR
julia> @btime intersect(1:1.0:4, 2:4)
  67.816 ns (2 allocations: 144 bytes)
3-element Vector{Float64}:
 2.0
 3.0
 4.0
```

IanButterworth commented:
Is it acceptable that the return type has changed in some of the examples above?

oscardssmith commented:
Imo, yes, but I'll put a triage label on it for good measure.

rfourquet commented:
> Is it acceptable that the return type has changed in some of the examples above?

Yes would be worth triaging, currently returning something of the same `eltype` as the first argument doesn't seem documented but might be a useful property? But if needed, it seems simple enough to update this PR to preserve this.

barucden commented:
Just a note: the return type has changed for the `AbstractRange`-`AbstractVector` version too, and there's no promotion -- the vector's eltype determines the result eltype. From Ian's table:

a | b | Master return | PR Return
-- | -- | -- | --
1:1.0:4 | [2,3,4,5,6] | [2.0, 3.0, 4.0] | [2, 3, 4]

Someone might dislike this inconsistency (promotion & no promotion) between the `AbstractRange`-`AbstractRange` and `AbstractRange`-`AbstractVector` version.

JeffBezanson commented on base/array.jl:
```diff
@@ -2612,3 +2612,10 @@ end
 setdiff(  itr, itrs...) = _shrink(setdiff!, itr, itrs)
+
+function intersect(v::AbstractVector, r::AbstractRange)
+    common = Iterators.filter(x -> x ∈ r, v)
```
```suggestion
    common = Iterators.filter(in(r), v)
```

JeffBezanson commented on base/array.jl:
```diff
@@ -2612,3 +2612,10 @@ end
+
+function intersect(v::AbstractVector, r::AbstractRange)
+    common = Iterators.filter(x -> x ∈ r, v)
+    seen = Set{eltype(v)}(common)
```
Is there a reason to do it this way instead of just `filter(in(r), v)`?

JeffBezanson replied:
Ah I see, it's to get unique values.

barucden replied:
Yes, `seen` contains unique elements. It is then used in the `_shrink_filter!` below, therefore the name `seen`. Should I put a short comment there?

JeffBezanson commented:
From triage: :+1: this is good. We think all the methods (of intersect and union) should be consistent with return type, so they should all promote.

IanButterworth commented:
Great. Looks like `union` is already consistent with promoting, from what I can tell
```julia
union(a,b)
```

| a            | b               | Return type     | Return                                              | `@btime`                               |
|--------------|-----------------|-----------------|-----------------------------------------------------|----------------------------------------|
| 1:4          | 2:4             | Vector{Int64}   | [1, 2, 3, 4]                                        | 227.164 ns (6 allocations: 544 bytes)  |
| 1:4          | 2.0:1.0:4.0     | Vector{Float64} | [1.0, 2.0, 3.0, 4.0]                                | 388.405 ns (11 allocations: 816 bytes) |
| 1:4          | 14:4:30         | Vector{Int64}   | [1, 2, 3, 4, 14, 18, 22, 26, 30]                    | 421.236 ns (12 allocations: 1.03 KiB)  |
| 1:4          | [2, 3, 4, 5, 6] | Vector{Int64}   | [1, 2, 3, 4, 5, 6]                                  | 346.486 ns (10 allocations: 656 bytes) |
| 1.0:1.0:4.0  | 2:4             | Vector{Float64} | [1.0, 2.0, 3.0, 4.0]                                | 383.856 ns (11 allocations: 816 bytes) |
| 1.0:1.0:4.0  | 2.0:1.0:4.0     | Vector{Float64} | [1.0, 2.0, 3.0, 4.0]                                | 289.467 ns (6 allocations: 544 bytes)  |
| 1.0:1.0:4.0  | 14:4:30         | Vector{Float64} | [1.0, 2.0, 3.0, 4.0, 14.0, 18.0, 22.0, 26.0, 30.0]  | 504.523 ns (12 allocations: 1.12 KiB)  |
| 1.0:1.0:4.0  | [2, 3, 4, 5, 6] | Vector{Float64} | [1.0, 2.0, 3.0, 4.0, 5.0, 6.0]                      | 420.648 ns (10 allocations: 752 bytes) |
| 10:4:30      | 2:4             | Vector{Int64}   | [10, 14, 18, 22, 26, 30, 2, 3, 4]                   | 427.095 ns (12 allocations: 1.03 KiB)  |
| 10:4:30      | 2.0:1.0:4.0     | Vector{Float64} | [10.0, 14.0, 18.0, 22.0, 26.0, 30.0, 2.0, 3.0, 4.0] | 517.068 ns (12 allocations: 1.12 KiB)  |
| 10:4:30      | 14:4:30         | Vector{Int64}   | [10, 14, 18, 22, 26, 30]                            | 262.619 ns (6 allocations: 544 bytes)  |
| 10:4:30      | [2, 3, 4, 5, 6] | Vector{Int64}   | [10, 14, 18, 22, 26, 30, 2, 3, 4, 5, 6]             | 578.182 ns (14 allocations: 1.73 KiB)  |
| [1, 2, 3, 4] | 2:4             | Vector{Int64}   | [1, 2, 3, 4]                                        | 332.230 ns (10 allocations: 656 bytes) |
| [1, 2, 3, 4] | 2.0:1.0:4.0     | Vector{Float64} | [1.0, 2.0, 3.0, 4.0]                                | 382.483 ns (10 allocations: 752 bytes) |
| [1, 2, 3, 4] | 14:4:30         | Vector{Int64}   | [1, 2, 3, 4, 14, 18, 22, 26, 30]                    | 418.221 ns (11 allocations: 1.00 KiB)  |
| [1, 2, 3, 4] | [2, 3, 4, 5, 6] | Vector{Int64}   | [1, 2, 3, 4, 5, 6]                                  | 260.780 ns (6 allocations: 544 bytes)  |

oscardssmith commented on test/arrayops.jl:
```diff
@@ -1106,6 +1106,9 @@ end
     @test isequal(intersect(Int64[]), Int64[])
+    @test isequal(intersect([1, 3], 1:typemax(Int)), [1, 3])
+    @test isequal(intersect(1:typemax(Int), [1, 3]), [1, 3])
+
```
Can we add a test or 2 for mixed types?

barucden replied:
Yes, check b77fc1c91211a16f204b47b4acd98b15e044c2ec please

barucden commented:
I found this case that does not promote:

```julia
julia> intersect(Set([1, 2, 3]), Set([1., 2., 3.]))
Set{Int64} with 3 elements:
  2
  3
  1

julia> intersect(Set([1., 2., 3.]), Set([1, 2, 3]))
Set{Float64} with 3 elements:
  2.0
  3.0
  1.0
```

Should it be addressed in this PR?

barucden commented:
NEWS entry and `compat` note added

oscardssmith commented:
> I found this case that does not promote:
> 
> ```julia
> julia> intersect(Set([1, 2, 3]), Set([1., 2., 3.]))
> Set{Int64} with 3 elements:
>   2
>   3
>   1
> 
> julia> intersect(Set([1., 2., 3.]), Set([1, 2, 3]))
> Set{Float64} with 3 elements:
>   2.0
>   3.0
>   1.0
> ```
> 
> Should it be addressed in this PR?

Yes.

rfourquet commented:
> We think all the methods (of intersect and union) should be consistent with return type, so they should all promote.

Is this meant to apply to `setdiff` and `symdiff` too? 
Given that the first argument is already special in these operations, in that it gives the type of container of the result, it would seem fine to me that the first argument also decide the `eltype` for operations which filter elements out, i.e. `intersect` and `setdiff`. I find this even easier to predict the resulting `eltype` than with promotion.

barucden commented:
I had to remove two `@test`s. They were in contradiction with the conclusion from triage. I think it's ok now except for the multi-argument case:

```julia
julia> intersect(Set([1,2,3]), [2, 3, 4], [2.0, 3.0])
Set{Int64} with 2 elements:
  2
  3

julia> union(Set([1,2,3]), [2, 3, 4], [2.0, 3.0])
Set{Float64} with 4 elements:
  4.0
  2.0
  3.0
  1.0
```

Union works correctly because it first creates an empty set with a promoted eltype, and then pushes in all the elements. For intersect ([this line](https://github.com/JuliaLang/julia/pull/41769/files#diff-bfc7bebef1dfd27f8485582c531c7cceb25b3aea8fcab3e4c2d0489964eecbe8R130)), I could think of something like this
```julia
function intersect(s::AbstractSet, itrs...)
    T = promote_eltype(s, itrs...)
    return intersect!(Set{T}(s), itrs...)
end
```
but I am wondering if there's something better?

IanButterworth commented:
It would be good to merge this. Perhaps your suggestion @barucden could be the starting point, so worth adding to see if it passes

barucden commented:
I added the multi-argument case and rebased on master.

IanButterworth commented:
The FreeBSD failure appears unrelated

barucden commented:
Thank you all for your help.

IanButterworth commented:
No, thank you @barucden! 👍🙂

vtjnash commented:
Seems like this might have destroyed performance of `intersect` on a BitSet: https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_date/2021-09/03/report.md. I think because we used to know that the intersection of a BitSet and anything else must fit into a BitSet, so we would prefer the narrower type (since we know the type-intersection is always sufficient), while now we copy it into a Set (i.e. always preferring the wider type).

Would you be willing to look into that?

barucden commented:
If my understanding is correct, you propose to make a special method `intersect(s1::BitSet, s2)` and make it return `BitSet` instead of `Set{promote_type(eltype(s1), eltype(s2))}`. I think this is related to what @rfourquet mentioned.

Wouldn't it, however, violate the decision 

>We think all the methods (of intersect and union) should be consistent with return type, so they should all promote.

from triage?

dawgwe1 commented on base/abstractset.jl:
```diff
@@ -45,8 +45,6 @@ Set{Int64} with 3 elements:
 """
 function union end
```
End product

dawgwe1 reviewed (commented):
Closed accounts

IanButterworth commented:
Seems like this needs some more triage given https://github.com/JuliaLang/julia/pull/41769#issuecomment-916139967

JeffBezanson commented on base/abstractset.jl:
```diff
@@ -125,9 +127,12 @@ Set{Int64} with 1 element:
-intersect(s::AbstractSet, itr, itrs...) = intersect!(intersect(s, itr), itrs...)
+function intersect(s::AbstractSet, itr, itrs...)
+    T = promote_eltype(s, itr, itrs...)
+    return intersect!(Set{T}(s), itr, itrs...)
```
This is a bug --- it means we give a different container type for 3 arguments than 2 (and a different type than 1.7).
On 1.7:
```
julia> intersect(BitSet([1,2]), [1,2], [2])
BitSet with 1 element:
  2

julia> intersect(BitSet([1,2]), [1,2])
BitSet with 2 elements:
  1
  2
```
on master:
```
julia> intersect(BitSet([1,2]), [1,2], [2])
Set{Int64} with 1 element:
  2

julia> intersect(BitSet([1,2]), [1,2])
BitSet with 2 elements:
  1
  2
```
